oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/core/expression/OAuth2AuthenticationRule.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2002-2018 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.core.expression;
+
+import org.junit.rules.MethodRule;
+import org.junit.runners.model.FrameworkMethod;
+import org.junit.runners.model.Statement;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.GrantedAuthority;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.oauth2.core.accesstoken.AccessTokenAuthority;
+import org.springframework.security.oauth2.resourceserver.authentication.OAuth2ResourceAuthenticationToken;
+
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * @author Josh Cummings
+ */
+public class OAuth2AuthenticationRule implements MethodRule {
```

Comment:
I haven't seen any tests in Spring Security that use `MethodRule`. We typically use Mockito to mock an `Authentication` or simply create an instance of `OAuth2ResourceAuthenticationToken` with the attributes required for the test.

Reply:
I like using a `Rule` here because it makes it easy to make sure the context gets cleaned up at the end.

I think that `TestRule` would be a better fit now that I know more about it, but initially, I was just following what I saw in `SpringTestRule`.

Either way, you'll notice that the rule does what you said: Creates an instance of `OAuth2ResourceAuthenticationToken`.

Of course, I can just create a helper class to do that for me, and I don't need a rule. But a `Rule` here does abstract away the clearing out of the security context, like `SpringTestRule` does with the `ApplicationContext`.

The alternative (which I'm fine with) is to use an @After to get the context cleaned up.

Reply:
My vote is to use `@After`. `SpringTestRule` is really nice and convenient and it does a lot more than closing the `ApplicationContext`.

Reply:
Actually, one more data point. I didn't realize this until this morning, but it looks like Rules are not natively supported in JUnit 5.

Given that, I'll change it to something JUnit 5 compatible.

Regarding your point, agreed--I see a common motive between the two rules, not that they are comparable in a wider sense.